rfc/core language/1461-cache-reservable.md

```diff
@@ -0,0 +1,200 @@
+# [RFC] Cache Reservation API
+
+**Authors:**
+
+- @hamersaw 
+
+## 1 Executive Summary
+
+The cache reservation API provides a lease-based reservation system for cachable tasks in Flyte. The purpose is to mitigate simultaneous evaluations of cachable tasks over identical inputs, resulting in duplication of work and therefore inefficient resource utilization. The proposed approach will more effectively process workflows with potentially significant improvements in end-to-end workflow evaluation times for instances with long running cachable tasks.
+
+## 2 Motivation
+
+Currently, Flyte initializes cachable tasks with a lookup the the datacatalog cache. If a previous instance of the task (ie. identical version and inputs) has completed the cached values are used, otherwise the task is executed.
+
+The issue is that disparate workflows, or unique executions of the same workflow, may execute an identical cachable task before a previous has completed. This results in multiple instances of the same task execution being performed simultaneously. For example, two workflows, namely A and B, contain the same long running (ex. 2 time units) identical cachable task. Workflow A executes the task beginning at t0 (finishing at t2) and workflow B executes the task at t1 (finishing at t3). The inefficiencies are twofold:
+
+1. From t1 (workflow B task execution) to t2 (workflow A task completion) there are two instances of the same task performing identical work (albeit at different stages).
+2. The execution from workflow B will not complete until t3, whereas it could use the cached results from workflow A at t2 to complete faster.
+
+The proposed solution will mitigate unnecessary resource utilization by disallowing duplicate task executions and provide more efficient workflow processing by using all available cachable task results.
+
+
+## Proposed Implementation
+
+### Guide-level explanation
+
+User-side functionality will require the introduction of a cacheReservation Flyte task annotation. An elementary example is how this may look is provided below:
+    
+```python
+@task(cacheReservation=true)
```

Comment:
@wild-endeavor @kumare3 Can you give your opinion about this flag?

```suggestion
@task(cache_serialized=true)
```
maybe?

Reply:
this was actually one of the slight differences. the flytekit integration is currently using cache_reservable to induce the API. the flag is used in tandem with the cache flag or it doesn't do anything. so tasks with this functionality enabled will be annotated as follows:

`@task(cache=True, cache_reservable=True, cache_version="1.0")`

this seemed a fairly idiomatic, introducing a layer on top of the existing cache scheme, but very open to feedback. can certainly make changes in the flytekit PR (once it's ready).
cc @eapolinario

Reply:
Ya I don't like cacheReservation. Cache_serialized=True is better.

Is there a timeout?

Reply:
Sounds good, I'll make the changes. Do we want to require the cache flag to be set as well? Or is setting cache_serialized enough to enable caching?

Is the timeout you mention for executions waiting on the cache to be populated? Is introducing another parameter the best solution here? The existing timeout parameter will cover the single executing task but not others that are waiting.

Reply:
Is there any advantage to generalizing this more? We can change the parameter to just "serialized" which means that only a single instance of the task can run at a time. The backend implementations will be exactly the same as they are now, just change the naming to better clarify what is happening.

`@task(serialized=True)`

And the obvious use case is for cachable tasks

`@task(cache=True, serialized=True)`

And can add a timeout

`@task(serialized=True, serialize_timeout=30m)`